scalableminds/webknossos
# Transactional saving of volume annotations

### Description:
- Introduces sensible transaction points for volume annotations (--> also means less versions which will be helpful in the version restore view :tada:). This is done by waiting for small time windows (1 second) in which no volume data was mutated **and** no dirty volume buckets need to be downloaded. This is similar to how the PushQueue operated, anyway, but it chunked buckets into 32 batches later which destroyed any transactional semantics. Reusing the state machine of the undo mechanism would be nicer, since it is strictly user-interaction based. However, the implementation of that would be far more complicated.
- Improved speed of bucket compression by using larger payloads for the webworkers (an isolated benchmark with reasonable inputs showed that the bucket compression is now 5 to 10 times faster).
- Removes AsyncTaskQueue as its sequentiality and failure-mechanism aren't needed anymore. [This abstraction was originally introduced to deal with failing network requests.](https://github.com/scalableminds/webknossos/blob/65319c4de0bb0193af4a47fc1ec746435395e9f9/app/assets/javascripts/oxalis/model/binary/layers/wk_layer.js#L96) However, that part is not handled by the AsyncTaskQueue anymore.
- The error handling in the PushQueue was hardened and unified with the error handling in the sagas (by escalating potential exceptions to the sagas).
- Fixes a small bug in the save-saga which could lead to an accumulating backlog of save queue items that would grow faster than they would be saved. This could happen when brushing a high volume of data and only relying on the auto-save mechanism. That mechanism did only send 1 request every 30 seconds (even when multiple requests would be necessary to completely save the state of that time).
- The save button now also warns when the PushQueue is starving (i.e., constantly unable to bundle up a transaction, because a bucket was edited again). Previously, this was achieved by providing a max_time parameter to _.debounce. However, since we want sensible transactions, this is not an option anymore.
- Move some modules from libs to libs/async.
- Lower saving batch constants for volume tracings, because the backend needs more time to process the update actions (see discussion [here](https://scm.slack.com/archives/C5AKLAV0B/p1692177165085429))
- Improve progress indicator during saving (the tooltip for the save button will also show when buckets need to be downloaded during saving, because this can take a noticable amount of time).

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- do a volume annotation (with and without fallback)
- set the network speed to very slow
- reload the tab during save (large volume changes are necessary to cause multiple requests)
- ensure that the volume data looks consistent (zoom in and out to check all mags)
- additionally/alternatively:
  - brush a large area
  - save
  - look at the version restore view. there should be one version (maybe other versions for the new segment id and for updateTracing, but not more)
  - compare the above to master.webknossos.xyz where lots of versions will be created. previewing one version and zooming to different mags will show hard-cuts in the brushed data.

### TODOs:
- [X] test volume speed on dev instance (which will use prod mode) to ensure that it didn't regress
- [X] verify that new save requests don't get too large. since updateBucket actions are way larger than skeleton update actions, some tuning might be necessary here. <-- no problem, since the compression does a good job of keeping the size small

### Issues:
- hopefully fixes [#7256](https://github.com/scalableminds/webknossos/issues/7256)
- contributes to #6495
- to close the issue fully, the version history view should show the transactions (the back-end needs to store these first, though)

------
(Please delete unneeded items, merge only when none are left open)
